### ISSUE TITLE
gRPC

### ISSUE THREAD

**Issue by moshababo** [ExchangeUnion/xud#33]
We want to switch our RPC layer to use [gRPC](https://grpc.io/).

At first, our current [RPC server ](https://github.com/ExchangeUnion/xud/blob/master/lib/rpc/RpcServer.js) need to be migrated, together with [XUClient](https://github.com/ExchangeUnion/xud/blob/master/lib/xuclient/XUClient.js) which consumes it. 

Second, the new gRPC server need to become web-compatible, to support browser clients. To do so, [grpcwebproxy](https://github.com/improbable-eng/grpc-web/tree/master/go/grpcwebproxy) or a similar like solution need to be integrated, in a way that it's activation will be configurable. To test it, a new separated web client repo need to be created.

**Comment by kilrau** [comment]
Research from @sangaman @moshababo to make a decision this week.

**Comment by itsbalamurali** [comment]
@moshababo  @sangaman @kilrau 

I've started working on protobuf's & GRPC,

We I thing can use: https://www.npmjs.com/package/grpc-dynamic-gateway to enable dynamic reverse proxy using google api annotations.

here is the WIP: 
```protobuf
syntax = "proto3";
//import "google/api/annotations.proto";
package xudrpc;

service XUDService {

 rpc GetInfo(GetInfoRequest) returns (GetInfoResponse) {
    option (google.api.http) = {
        get: "/v1/getinfo"
    };
  }

 rpc GetPairs(GetPairsRequest) returns (GetPairsResponse) {
    option (google.api.http) = {
        get: "/v1/getpairs"
    };
  }

  rpc GetOrders(OrdersRequest) returns (OrdersResponse) {
    option (google.api.http) = {
        get: "/v1/orders"
    };
  }

  rpc PlaceOrder(OrderRequest) returns (OrderResponse) {
    option (google.api.http) = {
        post: "/v1/orders"
        body: "*"
    };
  }

 rpc Connect(ConnectRequest) returns (ConnectResponse) {
    option (google.api.http) = {
        post: "/v1/connect"
        body: "*"
    };
  }

  rpc TokenSwap(TokenSwapRequest) returns (TokenSwapResponse) {
    option (google.api.http) = {
        post: "/v1/tokenswap"
        body: "*"
    };
  }
}

message GetInfoRequest {}
message GetInfoResponse {
  /// The identity pubkey of the current node.
  string identity_pubkey = 1 [json_name = "identity_pubkey"];
  /// If applicable, the alias of the current node, e.g. "bob"
  string alias = 2 [json_name = "alias"];
  /// Number of pending channels
  uint32 num_pending_channels = 3 [json_name = "num_pending_channels"];
  /// Number of active channels
  uint32 num_active_channels = 4 [json_name = "num_active_channels"];
  /// Number of peers
  uint32 num_peers = 5 [json_name = "num_peers"];
  /// The node's current view of the height of the best block
  uint32 block_height = 6 [json_name = "block_height"];
  /// The node's current view of the hash of the best block
  string block_hash = 8 [json_name = "block_hash"];
  /// Whether the wallet's view is synced to the main chain
  bool synced_to_chain = 9 [json_name = "synced_to_chain"];
  /// Whether the current node is connected to testnet
  bool testnet = 10 [json_name = "testnet"];
  /// A list of active chains the node is connected to
  repeated string chains = 11 [json_name = "chains"];
  /// The URIs of the current node.
  repeated string uris = 12 [json_name = "uris"];
}

message OrderRequest {}
message OrderResponse {}
message GetPairsRequest {}
message GetPairsResponse {}
message TokenSwapRequest {}
message TokenSwapResponse {}
message ConnectRequest {
  string host = 1;
  uint32 port = 2;
}
message ConnectResponse {}
message OrdersRequest {}
message OrdersResponse {}
```

**Comment by moshababo** [comment]
@itsbalamurali 
Looks fine for now.

Please elaborate more on your plans for `grpc-dynamic-gateway`.

**Comment by itsbalamurali** [comment]
@moshababo  please have a look at https://github.com/ExchangeUnion/xud/blob/grpc/lib/grpc/GrpcWebProxy.ts

**Comment by itsbalamurali** [comment]
yes, `grpcwebproxy` is written in golang and does the same thing. Hence I thought having the proxy inbuilt would make more sense having the option to either enable or disable API.

**Comment by moshababo** [comment]
Sounds good.

**Comment by kilrau** [comment]
Let's list missing gRPC methods like `rpc cancelOrder()` here and include them in this PR @sangaman @itsbalamurali

**Comment by sangaman** [comment]
Resolved with #103.
